keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+semantic models. This way, each model can focus on one aspect, like counting
+items or selection via labels.
+
+Workloads make claims to request resources with claim parameters that use CRDs
+provided together with a DRA driver. As soon as such a custom CR is created, a
+CRD controller of the DRA driver must convert it into a in-tree claim parameter
+object that uses the "parameter types" defined by the semantic models.
+
+The scheduler receives both the node resource information and the claim
+parameters. It uses the in-tree claim parameters to identify and allocate
+suitable resources. It captures all information that the DRA driver kubelet
+plugin will need in the "allocation types" of the semantic models and stores
+that in the claim status. Then it schedules the pod.
+
+The DRA driver kubelet plugin receives that information when the pod is about to
+start. It checks that the hardware is still available and healthy, then
+configures it for the pod as specified in the allocation types.
+
+### Publishing node resources
+
+A semantic model defines an attributes type. This is a Go type that gets
+embedded inside the resource.k8s.io API, not a separate API kind. The kubelet
+queries the registered DRA driver kubelet plugins for resource information
+through a new gRPC call. A streaming gRPC API allows drivers to update the
+capacity information.
+
+For example, a simple counter could be described in a `counter` Go package as:
+```
+type Attributes struct {
+    // Some meta information just for logging and debugging. Not used for counting.
+	metav1.ObjectMeta
+
+	Count int64
+}
+```
+
+The kubelet then publishes this information in a new NodeResources object on
+behalf of the driver. This implies that kubelet, DRA driver, and apiserver must
+all support the same semantic models and the same versions of them. It might be
+possible to support version skew (= keeping kubelet at an older version than
+the control plane and the DRA drivers) in the future, but currently this is out
+of scope.
+
+### Using semantic parameters as claim parameters
+
+The semantic model also defines a type for claim parameters. For the example
+above, that could be:
+```
+type Parameters struct {
+	Required int64
+}
+```
+
+The `Parameters.Required` value gets compared against `Attributes.Count`.
+
+Suppose a vendor has a CRD which allows requesting a certain number of `frobnicators`:
+
+```
+kind: ClaimParameters
+apiVersion: dra.frobincatorInc.example.com/v1alpha1
+metadata:
+  name: my-claim-parameters
+  uid: abcd
+  resourceVersion: 123
+spec:
+  requiredFrobincators: 100
+```
+
+The vendor CRD controller needs to sync this CRs into one in-tree object:
+```
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+metadata:
+  name: frobnicator-claim-parameters
+  namespace: default
+generatedFrom:
+  kind: ClaimParameters
+  apiGroup: dra.frobincatorInc.example.com
+  name: my-claim-parameters
+  resourceVersion: 123
```

Comment:
I think you want `generation` here, not `resourceVersion`, so that updates to metadata and status subresource do not bump the number.

Reply:
Is the semantic of `generation` well-defined enough for this? What if (for whatever reason) the CRD labels and annotations do influence the generated ResourceClaimParameter?

Demanding that the `resourceVersion` matches feels safer to me, but I don't feel strongly about this.

Reply:
Are all types required to have a `generation`? Can we assume that all CRDs have it?